File: internal/models/agent.go
package models

import (
	"database/sql"
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// Agent status constants
const (
	AgentStatusPending  = "pending"
	AgentStatusActive   = "active"
	AgentStatusInactive = "inactive"
	AgentStatusError    = "error"
)

// CertificateInfo stores information about an agent's certificate
type CertificateInfo struct {
	SerialNumber string    `json:"serial_number"`
	IssuedAt     time.Time `json:"issued_at"`
	ExpiresAt    time.Time `json:"expires_at"`
}

// Agent represents a registered agent in the system
type Agent struct {
	ID              string          `json:"id"`
	Name            string          `json:"name"`
	Status          string          `json:"status"`
	LastError       sql.NullString  `json:"lastError"`
	LastSeen        time.Time       `json:"lastSeen"`
	LastHeartbeat   time.Time       `json:"lastHeartbeat"`
	Version         string          `json:"version"`
	Hardware        Hardware        `json:"hardware"`
	CreatedByID     uuid.UUID       `json:"createdById"`
	CreatedBy       *User           `json:"createdBy,omitempty"`
	Teams           []Team          `json:"teams,omitempty"`
	CreatedAt       time.Time       `json:"createdAt"`
	UpdatedAt       time.Time       `json:"updatedAt"`
	Certificate     sql.NullString  `json:"-"`
	PrivateKey      sql.NullString  `json:"-"`
	CertificateInfo CertificateInfo `json:"certificate_info"`
}

// Hardware represents the hardware configuration of an agent
type Hardware struct {
	CPUs              []CPU              `json:"cpus"`
	GPUs              []GPU              `json:"gpus"`
	NetworkInterfaces []NetworkInterface `json:"network_interfaces"`
}

// CPU represents a CPU in the agent's hardware
type CPU struct {
	Model   string `json:"model"`
	Cores   int    `json:"cores"`
	Threads int    `json:"threads"`
}

// GPU represents a GPU in the agent's hardware
type GPU struct {
	Model  string `json:"model"`
	Memory string `json:"memory"`
	Driver string `json:"driver"`
}

// NetworkInterface represents a network interface in the agent's hardware
type NetworkInterface struct {
	Name      string `json:"name"`
	IPAddress string `json:"ip_address"`
}

// AgentMetrics represents real-time metrics from an agent
type AgentMetrics struct {
	AgentID     string    `json:"agent_id"`
	CPUUsage    float64   `json:"cpu_usage"`
	GPUUsage    float64   `json:"gpu_usage"`
	GPUTemp     float64   `json:"gpu_temp"`
	MemoryUsage float64   `json:"memory_usage"`
	Timestamp   time.Time `json:"timestamp"`
}

// ScanHardware scans a JSON-encoded hardware string into the Hardware struct
func (a *Agent) ScanHardware(value interface{}) error {
	if value == nil {
		a.Hardware = Hardware{}
		return nil
	}

	switch v := value.(type) {
	case []byte:
		return json.Unmarshal(v, &a.Hardware)
	case string:
		return json.Unmarshal([]byte(v), &a.Hardware)
	default:
		return fmt.Errorf("unsupported type for hardware: %T", value)
	}
}

// Value returns the JSON encoding of Hardware for database storage
func (h Hardware) Value() (driver.Value, error) {
	return json.Marshal(h)
}

// Scan implements the sql.Scanner interface for CertificateInfo
func (ci *CertificateInfo) Scan(value interface{}) error {
	if value == nil {
		*ci = CertificateInfo{}
		return nil
	}

	switch v := value.(type) {
	case []byte:
		return json.Unmarshal(v, ci)
	case string:
		return json.Unmarshal([]byte(v), ci)
	default:
		return fmt.Errorf("unsupported type for certificate_info: %T", value)
	}
}

// Value implements the driver.Valuer interface for CertificateInfo
func (ci CertificateInfo) Value() (driver.Value, error) {
	return json.Marshal(ci)
}

File: frontend/src/types/agent.ts
/**
 * Agent types and interfaces for the KrakenHashes frontend.
 * 
 * @packageDocumentation
 */

import { JobTask, JobExecution } from './jobs';

/**
 * Represents a registered agent in the system.
 * 
 * @interface Agent
 * @property {string} id - Unique identifier for the agent (UUID)
 * @property {string} name - Display name of the agent
 * @property {'inactive' | 'active' | 'error'} status - Current agent status
 * @property {string} lastHeartbeat - ISO timestamp of last heartbeat
 * @property {number} createdBy - User ID of agent creator
 * @property {string} createdAt - ISO timestamp of creation
 * @property {string} updatedAt - ISO timestamp of last update
 * @property {AgentMetrics} [metrics] - Optional current metrics
 */
export interface Agent {
    id: string;
    name: string;
    status: 'inactive' | 'active' | 'error';
    lastHeartbeat: string;
    createdBy: {
        id: string;
        username: string;
    };
    version: string;
    hardware: AgentHardware;
    teams: {
        id: string;
        name: string;
    }[];
    createdAt: string;
    updatedAt: string;
    metrics?: AgentMetrics;
    isEnabled?: boolean;
    ownerId?: string;
    extraParameters?: string;
    metadata?: {
        busy_status?: string;
        current_task_id?: string;
        current_job_id?: string;
        [key: string]: any;
    };
}

/**
 * Represents an agent with its current task information.
 * Used in the dashboard to show agents and their active jobs.
 * 
 * @interface AgentWithTask
 * @extends Agent
 * @property {JobTask} [currentTask] - The current task assigned to this agent
 * @property {JobExecution} [jobExecution] - The job execution for the current task
 */
export interface AgentWithTask extends Agent {
    currentTask?: JobTask;
    jobExecution?: JobExecution;
}

/**
 * Represents real-time metrics data from an agent.
 * 
 * @interface AgentMetrics
 * @property {number} cpuUsage - CPU usage percentage (0-100)
 * @property {number} gpuUsage - GPU usage percentage (0-100)
 * @property {number} gpuTemp - GPU temperature in Celsius
 * @property {number} memoryUsage - Memory usage percentage (0-100)
 * @property {string} timestamp - ISO timestamp of metrics collection
 */
export interface AgentMetrics {
    cpuUsage: number;
    gpuUsage: number;
    gpuTemp: number;
    memoryUsage: number;
    timestamp: string;
}

/**
 * Form data structure for registering a new agent.
 * 
 * @interface AgentRegistrationForm
 * @property {string} name - Desired name for the new agent
 * @property {number} teamId - Team ID to associate with the agent
 * @property {boolean} continuous - Whether the claim code can be used multiple times
 */
export interface AgentRegistrationForm {
    name: string;
    teamId: number;
    continuous: boolean;
}

/**
 * Structure for agent claim codes.
 * 
 * @interface ClaimCode
 * @property {string} code - The generated claim code
 * @property {boolean} continuous - Whether the code can be used multiple times
 * @property {string} createdAt - ISO timestamp of code generation
 */
export interface ClaimCode {
    code: string;
    continuous: boolean;
    createdAt: string;
}

/**
 * Represents hardware information reported by an agent
 */
export interface AgentHardware {
    cpus: {
        model: string;
        cores: number;
        threads: number;
    }[];
    gpus: {
        model: string;
        memory: string;
        driver: string;
    }[];
    networkInterfaces: {
        name: string;
        ipAddress: string;
    }[];
}

/**
 * Represents a device detected by an agent
 */
export interface AgentDevice {
    id: number;
    agent_id: number;
    device_id: number;
    device_name: string;
    device_type: string;
    enabled: boolean;
    created_at: string;
    updated_at: string;
}

/**
 * Represents a claim voucher in the system
 */
export interface ClaimVoucher {
    code: string;
    created_by: {
        id: string;
        username: string;
        email: string;
        role: string;
    };
    created_by_id: string;
    is_continuous: boolean;
    is_active: boolean;
    created_at: string;
    updated_at: string;
    used_at?: {
        Time: string;
        Valid: boolean;
    };
    used_by_agent_id?: {
        Int64: number;
        Valid: boolean;
    };
} 